### PR TITLE
Add warning for use of lifetime parameter with 'static bound

### LINKED ISSUE #40661: "use of undeclared lifetime name `'static`" is misleading
**Code:** https://is.gd/6L8tEl

```rust
fn foo<'a: 'static>() {}
```

**Output:**

```
rustc 1.17.0-nightly (b1e31766d 2017-03-03)
error[E0261]: use of undeclared lifetime name `'static`
 --> <anon>:1:12
  |
1 | fn foo<'a: 'static>() {}
  |            ^^^^^^^ undeclared lifetime
```

**Expectation:**

`<'a: 'static>` is unnecessary (you can just use `'static` in place of `'a`), but maybe we should accept it? At any rate, the error message is misleading - `'static` is the one globally declared lifetime. If we want to reject this code, the error message should not imply `'static` is undefined, but rather that you just don't need the `'a` lifetime.

### PR DESCRIPTION
**Pull request by adamransom**
Previously a `'static` lifetime bound would result in an `undeclared lifetime` error when compiling, even though it could be considered valid.

However, it is unnecessary to use it as a lifetime bound so we present the user with a warning instead and suggest using the `'static` lifetime directly, in place of the lifetime parameter. We can change this to an error (or warning with lint) if that's decided to be more appropriate.

Example output:
```
warning: unnecessary lifetime parameter `'a`
 --> ../static-lifetime-bound.rs:3:10
  |
3 | fn f<'a: 'static>(val: &'a i32) {
  |      ^^^^^^^^^^^
  |
  = help: you can use the `'static` lifetime directly, in place `'a`
```

Fixes #40661

r? @jseyfried

### CODE REVIEW

File: src/librustc/middle/resolve_lifetime.rs [hunk 1]
```diff
@@ -434,11 +434,11 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
             self.resolve_elided_lifetimes(slice::ref_slice(lifetime_ref));
             return;
         }
-        if lifetime_ref.name == keywords::StaticLifetime.name() {
+        if lifetime_ref.is_static() {
             self.insert_lifetime(lifetime_ref, Region::Static);
             return;
         }
-        self.resolve_lifetime_ref(lifetime_ref);
+        self.resolve_lifetime_ref(lifetime_ref, lifetime_ref);
```
Comment (jseyfried) [inline src/librustc/middle/resolve_lifetime.rs]: I would make the second parameter `Option<&hir::Lifetime>` and use `None` here to make it clear that the second argument won't get used.

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -917,6 +917,16 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                 }
             }
             self.insert_lifetime(lifetime_ref, def);
+        } else if lifetime_ref.is_static() {
+            // The only way we end up here is if there is a static lifetime bound (due to early
+            // return on `visit_lifetime()`)
+            self.insert_lifetime(lifetime_ref, Region::Static);
+            let full_span = mk_sp(context_ref.span.lo, lifetime_ref.span.hi);
```
Comment (jseyfried) [inline src/librustc/middle/resolve_lifetime.rs]: This would become `context_ref.unwrap().span.lo`, enforcing the invariant described in the above comment.

File: src/librustc/middle/resolve_lifetime.rs [hunk 3]
```diff
@@ -917,6 +917,16 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
+            self.sess.struct_span_warn(full_span,
+                &format!("unnecessary lifetime parameter `{}`", context_ref.name))
+                .help(&format!("you can use a `'static` lifetime directly, in place of \
+                                the lifetime parameter `{}`", context_ref.name))
```
Comment (jseyfried) [inline src/librustc/middle/resolve_lifetime.rs]: I would reword: ``you can use the lifetime `'static` directly in place of `{}` ``.

**Review by jseyfried — COMMENTED** [review]
Looks good!

**Comment by jseyfried** [comment]
Nominating for @rust-lang/lang decision: should we allow, warn, or forbid a `'static` bound on a lifetime parameter, e.g. `fn f<'a: 'static>(_: &'a ()) {}`?

Today, it causes a [confusing error](https://play.rust-lang.org/?gist=bf03d7ebedfbe3c7d1d7f3cb4b2fd67d&version=nightly&backtrace=0).

File: src/librustc/middle/resolve_lifetime.rs (hunk 3 above)
Comment (adamransom) [inline src/librustc/middle/resolve_lifetime.rs]: Yeah, that's better. I was also wondering if it's worth adding a note? Something like:
```
note: due to use of `'static` bound on `'a`
```
or whether the help message alone is enough?

File: src/librustc/middle/resolve_lifetime.rs (hunk 1 above)
Comment (adamransom) [inline src/librustc/middle/resolve_lifetime.rs]: Yes, I was thinking about this also. I'll add that.

File: src/librustc/middle/resolve_lifetime.rs (hunk 3 above)
Comment (jseyfried) [inline src/librustc/middle/resolve_lifetime.rs]: The help message is sufficient, imo.

**Comment by nikomatsakis** [comment]
@jseyfried conclusion from @rust-lang/lang meeting was that we should accept it, but warn (as I believe this PR does)

File: src/librustc/middle/resolve_lifetime.rs [hunk 4]
```diff
@@ -860,7 +860,9 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
         });
     }
 
-    fn resolve_lifetime_ref(&mut self, lifetime_ref: &hir::Lifetime) {
+    fn resolve_lifetime_ref(&mut self,
+                            lifetime_ref: &hir::Lifetime,
+                            context_ref: Option<&hir::Lifetime>) {
```
Comment (eddyb) [inline src/librustc/middle/resolve_lifetime.rs]: The terminology here is wrong IMO. The intention of `_ref`, AFAICT, is to mean "lifetime reference", as in, one use of  (referring to) a lifetime. So this extra argument should be `maybe_def` or some such, but even then, I would just bypass this whole function from `check_lifetime_defs`, for `'static`, i.e. handle it there instead of here.

Comment (adamransom) [inline src/librustc/middle/resolve_lifetime.rs]: Yes, I agree that's a better place to handle it. Will make that change!

**Comment by adamransom** [comment]
Updated with latest suggestions.

File: src/librustc/middle/resolve_lifetime.rs [hunk 5]
```diff
@@ -1464,7 +1464,17 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
             self.check_lifetime_def_for_shadowing(old_scope, &lifetime_i.lifetime);
 
             for bound in &lifetime_i.bounds {
-                self.resolve_lifetime_ref(bound);
+                if !bound.is_static() {
```
Comment (adamransom) [inline src/librustc/middle/resolve_lifetime.rs]: I chose to put the common path first, even though its a negated condition. I can change if this is not preferred style though.

**Comment by jseyfried** [comment]
@adamransom Thanks!
@bors r+

File: src/librustc/middle/resolve_lifetime.rs (hunk 5 above)
Comment (eddyb) [inline src/librustc/middle/resolve_lifetime.rs]: Looks good!

**Comment by arielb1** [comment]
- #40821
@bors retry

**Comment by arielb1** [comment]
-  #40821
    @bors retry

**Comment by arielb1** [comment]
-  #40821
    @bors retry
